### DISCUSSION TITLE
How is 'use displaced mesh' implemented?

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#24042] [Q&A Modules: Navier-Stokes] [answered]
Hello all,

Whilst experimenting with FSI in the FE Navier-Stokes module, I have encountered a situation where setting `use_displaced_mesh = true` for the Navier-Stokes kernels gives me different solutions when using the vector family of kernels, compared to the field variable family.

As in my other recent posts, the testing ground has been the flat channel example discussed [here](https://github.com/idaholab/moose/discussions/23484).

Running the original, and vector variable versions of this file with `use_displaced_mesh = false` everywhere produces identical results in both the solid and fluid domains. This is great news, and I believe indicates that the custom interface kernels (thanks again for the input [here](https://github.com/idaholab/moose/discussions/23658)!) are doing exactly the same thing as the original versions.

However, when I turn on `use_displaced_mesh`, the solutions become different. 

I am therefore wondering if the displaced mesh system should behave any differently with either vector variables, or for some reason specifically in the vector family of Navier-Stokes kernels? So far, I haven't found anything in the code that would suggest this...

Thanks!

**Comment by lindsayad** [comment, marked as answer]
Conceptually they should not behave differently; in practice ... well it seems like they do 😦

**Reply by lindsayad** [reply]
One thing I will say is that the vector kernels have built-in addition of derivatives with respect to the mesh displacements to the jacobian, whereas the scalar field kernels do not. However, that should only impact the convergence efficiency, and if both solves converge then the solutions should compare equal within a tolerance.

**Reply by Eilloo** [reply]
Interesting... Yes, I've seen some cases where the convergence is different depending on which kernels I turn on displaced mesh for. In fact, the interface diffusion kernels seem to have the biggest effect here, with the non-vector version failing to converge at all when displaced mesh is on (by default and in the example file though, it's off). Nice to know where that might be coming from.

If I uncover anything explaining the differences in the actual solutions, I'll keep you updated.

For these simple test cases with very few elements, the differences are noticeable, but pretty small. Going up to more complex cases though, these differences seem to compound resulting in wildly different behaviour between vector and non-vector stuff. (Of course, that's assuming this discrepancy is the cause, and not some other mistake 😆)

**Reply by Eilloo** [reply]
Following up on this, does anyone know where I can look for the code implementation of `use_displaced_mesh = true` compared to `false`? 
I can see that `displaced_problem` seems to be a whole system of its own, but I can't find anything vector or non-vector specific so far.
In the kernel base class, I can see the declaration of `use_displaced_mesh`, but not the code where things are calculated differently based on whether this is true or false.

Thanks!

**Reply by GiudGiud** [reply]
Hello

This will very much depend on the object type. 
For example for ADKernels:
```
ADKernelTempl<T>::computeResidual()
{
  precalculateResidual();

  std::vector<Real> residuals(_test.size(), 0);

  if (_use_displaced_mesh)
    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
      for (_i = 0; _i < _test.size(); _i++)
        residuals[_i] += raw_value(_ad_JxW[_qp] * _ad_coord[_qp] * computeQpResidual());
```
the difference here is the use of the local element jacobian/weights with AD derivatives with regards to the displacements of the mesh

Guillaume

**Reply by Eilloo** [reply]
Thanks Guillaume,
This does make me think, if the Jacobian is used in the `computeResidual()` method, does this mean that the addition of derivatives mentioned by Alex above could in fact impact the solution, rather than just the convergence? Or have I misunderstood this...

**Reply by GiudGiud** [reply]
The solution doesnt change, as the residual is the same. 
This of an AD number for the contribution to the residual as :  (residual, d_residual_variable_1_local_cell, d_residual_variable_2_local_cell ...)
only the first component gets used to compute the residual, then the derivatives are used to build the Jacobian

**Reply by lindsayad** [reply]
@Eilloo are you setting `use_displaced_mesh` on the INSAD material? This is essential to get the same results as when using the INS objects since the INSAD material computes the strong residuals that are used by most of the INSAD kernels, e.g. the INSAD objects rarely do computations themselves. I made this mistake myself yesterday

**Reply by Eilloo** [reply]
This is a very good point and no, I haven't been - I'll test that as soon as I can and confirm any improvement.
Thanks also for highlighting [24024](https://github.com/idaholab/moose/pull/24204) - I'll add a link here too in case anyone finds this thread in future

**Reply by GiudGiud** [reply]
@Eilloo is this resolved?

**Reply by Eilloo** [reply]
Hi yes, sorry - forgot to mark as answer
